cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -261,19 +283,29 @@ void optimize_layout(T *head_embedding, int head_n, T *tail_embedding,
   MLCommon::device_buffer<T> epoch_of_next_sample(d_alloc, stream, nnz);
   MLCommon::copy(epoch_of_next_sample.data(), epochs_per_sample, nnz, stream);
 
+  MLCommon::device_buffer<T> embedding_updates(d_alloc, stream, n_vertices * params->n_components);
```

Comment:
I'm on the fence about doing this as the default behavior vs having this be a separate path only when the user explicitly sets a random_state (which won't be often). For example, I was looking through the UMAP reference documentation and they seem to have the same problem [1]. 

I don't believe the extra kernel likely is going to add much to the end-to-end runtime, but I do think we should verify that to be sure. My main hang-up is really just with the extra memory usage, that will grow with larger inputs. It doesn't seem like it would be hard for us to make the synchronous gradient updates an optional code path. Perhaps we can provide a new option `synchronize_updates` for users who may still want to use a completely random seed but still want to guarantee consistency.

[1] https://umap-learn.readthedocs.io/en/latest/reproducibility.html

Reply:
Ok, so do you prefer the one or two pathway(s) solution?

Reply:
Please forgive the confusion. When we discussed this offline yesterday, I had not yet seen the documentation about reproducibility in the reference implementation. After seeing that, I think it's okay that we use the deterministic code path as an option, which is enabled only when the user manually sets `random_state`.

This way, cuml will continue to function with the same side-effects as the multi-core execution in the reference implementation, but provide a path to deterministic reproducibility when needed. 

Aside from reproducibility for pytests, we have a user interested in this feature so they can produce the exact same visualization at different times. Otherwise, I think keeping the data races that exist n the reference implementation, for the purposes of speed and lowered memory usage, is okay.